### PR TITLE
openssh: backport zlib-1.3 support

### PR DESCRIPTION
**Pull request by trofi**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/openssh/common.nix [hunk 1]
```diff
@@ -45,10 +45,14 @@ stdenv.mkDerivation {
...
-    # On Hydra this makes installation fail (sometimes?),
-    # and nix store doesn't allow such fancy permission bits anyway.
     ''
+      # On Hydra this makes installation fail (sometimes?),
+      # and nix store doesn't allow such fancy permission bits anyway.
       substituteInPlace Makefile.in --replace '$(INSTALL) -m 4711' '$(INSTALL) -m 0711'
+
+      # Apply zlib-1.3 fix:
+      #   https://github.com/openssh/openssh-portable/commit/cb4ed12ffc332d1f72d054ed92655b5f1c38f621
+      substituteInPlace configure --replace 'if (n != 3 && n != 4)' 'if (n < 1)'
```
Comment (risicle) [inline pkgs/tools/networking/openssh/common.nix:55]: Assume this is because we don't want to have to re-auto-gen everything.

Comment (trofi) [reply pkgs/tools/networking/openssh/common.nix]: I was on the fence thinking of using `autoreconfHook` or not for a small change. I can see why a raw replacement is confusing. Switched to `autoreconfHook`.

Comment (dasJ) [reply pkgs/tools/networking/openssh/common.nix]: This is only until the next version is released, or am I mistaken here?

**Review by dasJ — APPROVED** [review]

File: pkgs/tools/networking/openssh/common.nix (hunk 1 above)
Comment (trofi) [reply pkgs/tools/networking/openssh/common.nix]: Yes, next release will contain the fix.

**Pull request merged**